examples/networking/ping_reply.py

```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/env python
+"""
+ping_reply.py
+    - Hooks at Traffic Control (ingress queue) in Kernel using eBPF
+    - If packet is an ICMP echo (ping)
+        * Intercepts the packet at traffic control ingress queue
+        * Modifies ICMP packet type to ICMP reply
+        * Swaps src/dst ip and mac and updates chksum
+        * Redirects packet to the same interface as egress
+
+USAGE: sudo python ping_reply.py
+"""
+
+from bcc import BPF
+from pyroute2 import IPRoute, IPDB
+import sys
+
+prog = """
+  #include <bcc/proto.h>
+  #include <uapi/linux/if_ether.h>
+  #include <uapi/linux/in.h>
+  #include <uapi/linux/icmp.h>
+ 
+  #define IP_SRC_OFF 26
+  #define IP_DST_OFF 30
+  #define IP_CKSUM_OFF 24
+  /*
+  There are better ways to get OFFSETs for e.g. by using macros from .h
+  files of other networking subsytems
+  */
+
+  /*
+  struct icmp_t would come in handy for parsing ICMP packets,
+  since bcc/proto.h doesn't specify a struct for parsing ICMP packets
+  */
+  struct icmp_t {
+        unsigned char type;
+        unsigned char code;
+        unsigned short cksum;
+        /* Fields only valid for echo-reply ICMP message */
+        unsigned short id;
+        unsigned short seq;
+  } BPF_PACKET_HEADER;
+
+  /*
+  struct __sk_buff is passed by the kernel to this function
+  */
+  int ping_reply (struct __sk_buff * skb) {
+        // Packet is available on the 0th location of the memory.
+        u8 *cursor = 0;
+
+        struct ethernet_t * ethernet = cursor_advance(cursor,
+                sizeof(*ethernet));
+        // If packet type is not IP, return
+        if (ethernet->type != ETH_P_IP)
+                return 0;
+
+        struct ip_t * ip = cursor_advance(cursor, sizeof(*ip));
+        // If next protocol is not ICMP, return
+        if (ip->nextp != IPPROTO_ICMP)
+                return 0;
+
+        struct icmp_t * icmp = cursor_advance(cursor, sizeof(*icmp));
+        // If ICMP packet is not echo, return
+        if (icmp->type != ICMP_ECHO)
+                return 0;
+
+        /*
+        Converting ICMP echo into ICMP reply by changing the type to 0
+        Since we're changing packet contents, we need to update the checksum
+        */
+        unsigned short type = ICMP_ECHOREPLY;
+        bpf_l4_csum_replace(skb,36,icmp->type, type,sizeof(type));
```

Comment:
It would be better to use incr_cksum_l4 helper function from helpers.h, which invokes some logic inside clang to understands bitfields and network byte order better.